crates/provider/src/layers/cache.rs

```diff
@@ -0,0 +1,753 @@
+
+            if let Some(hash) = params_hash {
+                if let Ok(Some(cached)) = self.cache.get_deserialized(&hash) {
+                    return ProviderCall::BoxedFuture(Box::pin(async move { Ok(cached) }));
+                }
+            }
+        }
+
+        let client = self.inner.weak_client();
+        let cache = self.cache.clone();
+
+        ProviderCall::BoxedFuture(Box::pin(async move {
+            let client = client
+                .upgrade()
+                .ok_or_else(|| TransportErrorKind::custom_str("RPC client dropped"))?;
+
+            let result = client.request(req.method(), req.params()).await?;
+
+            let json_str = serde_json::to_string(&result).map_err(TransportErrorKind::custom)?;
+
+            if !redirect {
+                let hash = req.params_hash()?;
+                let _ = cache.put(hash, json_str);
+            }
+
+            Ok(result)
+        }))
+    }
+
+    fn get_proof(
+        &self,
+        address: Address,
+        keys: Vec<StorageKey>,
+    ) -> RpcWithBlock<T, (Address, Vec<StorageKey>), EIP1186AccountProofResponse> {
+        let client = self.inner.weak_client();
+        let cache = self.cache.clone();
+        RpcWithBlock::new_provider(move |block_id| {
+            let req =
+                RequestType::new("eth_getProof", (address, keys.clone())).with_block_id(block_id);
+            cache_rpc_call_with_block!(cache, client, req)
+        })
+    }
+
+    fn get_storage_at(
+        &self,
+        address: Address,
+        key: U256,
+    ) -> RpcWithBlock<T, (Address, U256), StorageValue> {
+        let client = self.inner.weak_client();
+        let cache = self.cache.clone();
+        RpcWithBlock::new_provider(move |block_id| {
+            let req = RequestType::new("eth_getStorageAt", (address, key)).with_block_id(block_id);
+            cache_rpc_call_with_block!(cache, client, req)
+        })
+    }
+
+    fn get_code_at(&self, address: Address) -> RpcWithBlock<T, Address, Bytes> {
+        let client = self.inner.weak_client();
+        let cache = self.cache.clone();
+        RpcWithBlock::new_provider(move |block_id| {
+            let req = RequestType::new("eth_getCode", address).with_block_id(block_id);
+            cache_rpc_call_with_block!(cache, client, req)
+        })
+    }
+
+    fn get_transaction_count(&self, address: Address) -> RpcWithBlock<T, Address, U64, u64> {
+        let client = self.inner.weak_client();
+        let cache = self.cache.clone();
+        RpcWithBlock::new_provider(move |block_id| {
+            let req = RequestType::new("eth_getTransactionCount", address).with_block_id(block_id);
+
+            cache_rpc_call_with_block!(cache, client, req)
+        })
+    }
+
+    async fn get_logs(&self, filter: &Filter) -> TransportResult<Vec<Log>> {
+        let req = RequestType::new("eth_getLogs", filter.clone());
+
+        let params_hash = req.params_hash().ok();
```

Comment:
this is also a bit nuanced because the blockrange variants allows the server to chose the range for example if upper bound is missing